File: apps/marching-arts/marching_arts/bands.py
"""Data-classification bands.

Every fact the platform stores is about a person, and every one carries a band.
The band is a column from the first migration, not a decoration added later:
you can add the column to a schema afterwards, but you cannot add it to data
that was written without it.

The bands are ordered by sensitivity, and the order is load-bearing — a grant
names the highest band it authorizes and covers everything below it. Do not
reorder these values; existing rows carry the integers.
"""
from __future__ import annotations

from enum import IntEnum


class Band(IntEnum):
    """L0 through L6, least to most sensitive."""

    SELF = 0             # the member's own view of their own record
    ROSTER = 1           # name, section, part assignment — the org chart
    CRAFT = 2            # rehearsal notes, technique, what a section leader teaches to
    ACCOMMODATION = 3    # what someone needs. NEVER forwarded as a fact — see below.
    HEALTH = 4           # medical. Named persons only, never roles.
    SAFEGUARDING = 5     # routed, never received. Nothing here is served by this app.
    FAMILY = 6           # family and financial circumstance


#: At and above this band the platform serves the *derived instruction* and never
#: the underlying fact. A section leader learns "rotate this member out every
#: twenty minutes"; they do not learn why. The mechanism is a projection in the
#: SELECT list (see policy.Policy.projection) rather than a promise in a doc.
DERIVE_AT = Band.ACCOMMODATION

#: Bands this application refuses to serve at all, to anyone, under any grant.
#: L5 is routed to the people whose job it is and never received here: in every
#: leadership-implicating case on the public record, surfacing was external, so
#: an intake would digitise a broken path rather than repair it.
NEVER_SERVED = frozenset({Band.SAFEGUARDING})


def parse(value: "int | str | Band") -> Band:
    """Coerce a stored integer or a name to a Band, rejecting anything else.

    Fails loudly. A band that cannot be resolved is not defaulted to SELF — a
    silent downgrade to the least sensitive value is exactly the wrong failure
    direction for this column.
    """
    if isinstance(value, Band):
        return value
    if isinstance(value, int):
        return Band(value)
    return Band[str(value).strip().upper()]

File: apps/marching-arts/browser/src/bands.ts
/*
 * Copyright 2026 The marching-arts Authors
 * SPDX-License-Identifier: Apache-2.0
 *
 * Data-classification bands. Port of marching_arts/bands.py.
 *
 * The integers are the contract: they are written into the `band` column from
 * migration 001 and existing rows carry them. Do not reorder. The ordering is
 * load-bearing in two places — a grant names the highest band it authorizes and
 * covers everything below it, and `DERIVE_AT` is a `>=` comparison.
 *
 * This file is a transcription, not a design. If it disagrees with bands.py the
 * differential suite fails, and bands.py is right.
 */

/** L0 through L6, least to most sensitive. Mirrors `bands.Band`. */
export const Band = {
  SELF: 0,
  ROSTER: 1,
  CRAFT: 2,
  ACCOMMODATION: 3,
  HEALTH: 4,
  SAFEGUARDING: 5,
  FAMILY: 6,
} as const;

export type Band = (typeof Band)[keyof typeof Band];

/** Names in declaration order, so `parseBand` can accept a name like Python's `Band[...]`. */
export const BAND_NAMES = Object.keys(Band) as (keyof typeof Band)[];

/** Lowest and highest band integers — the CHECK bounds in migration 001. */
export const BAND_MIN: number = Math.min(...Object.values(Band));
export const BAND_MAX: number = Math.max(...Object.values(Band));

/**
 * At and above this band the platform serves the *derived instruction* and never
 * the underlying fact. Enforced as a projection in the SELECT list, not as a
 * promise in a doc — see `policy.Policy.projection`.
 */
export const DERIVE_AT: Band = Band.ACCOMMODATION;

/**
 * Bands this application refuses to serve at all, to anyone, under any grant.
 * Compiled into a DENY rule, so it negates the *union* of the allows and no
 * grant can win against it.
 */
export const NEVER_SERVED: ReadonlySet<Band> = new Set<Band>([Band.SAFEGUARDING]);

/**
 * Coerce a stored integer or a name to a Band, rejecting anything else.
 *
 * Fails loudly, exactly as `bands.parse` does. A band that cannot be resolved is
 * *not* defaulted to SELF: a silent downgrade to the least sensitive value is
 * the wrong failure direction for this column.
 */
export function parseBand(value: number | string | Band): Band {
  if (typeof value === 'number') {
    if (!Number.isInteger(value) || !(Object.values(Band) as number[]).includes(value)) {
      throw new RangeError(`${value} is not a valid Band`);
    }
    return value as Band;
  }
  if (typeof value === 'string') {
    const key = value.trim().toUpperCase() as keyof typeof Band;
    if (!(key in Band)) throw new RangeError(`${JSON.stringify(value)} is not a valid Band name`);
    return Band[key];
  }
  throw new TypeError(`cannot parse ${typeof value} as a Band`);
}

/** The DENY fragment's band list, rendered the way policy.py renders it. */
export function neverServedList(): string {
  return [...NEVER_SERVED].sort((a, b) => a - b).map((b) => String(b)).join(', ');
}
